### PR TITLE
Allow conversions between compatible array types

### LINKED ISSUE #80586: Functions accept typed arrays and cannot inherit conversions
### Godot version

4.2 dev3

### System information

Win10 Vulkan

### Issue description

```gdscript
extends Node2D

class A:
    func say()->void:
        print('A')

class B extends A:
    func say()->void:
        print('B')
		
func test(a:A)->void:
    a.say()
	
func test_arr(a:Array[A])->void:
    for _a in a:
        _a.say()

func _ready() -> void:
    test(B.new())
    var _b:Array[B]=[B.new(),B.new()]
    # Implicit type conversion,run error
    test_arr(_b)
    # Explicit type conversion
    test_arr(_b as Array[A]) 
	
    #only stupid manual conversion
    var _a:Array[A]=[]
    for i in _b:
        _a.append(i)

    test_arr(_a)


```

### Steps to reproduce

see description

### Minimal reproduction project

None

### PR DESCRIPTION
**Pull request by garychia**
Fixes #80586

### CODE REVIEW

File: modules/gdscript/gdscript_function.h [hunk 1]
```diff
@@ -84,7 +84,7 @@ class GDScriptDataType {
 						Ref<Script> array_script_type_ref = array.get_typed_script();
 
 						if (array_script_type_ref.is_valid()) {
-							valid = (container_element_type->kind == SCRIPT || container_element_type->kind == GDSCRIPT) && container_element_type->script_type == array_script_type_ref.ptr();
+							valid = (container_element_type->kind == SCRIPT || container_element_type->kind == GDSCRIPT) && container_element_type->script_type->cast_to<Script>(array_script_type_ref.ptr());
```
Comment (akien-mga) [inline modules/gdscript/gdscript_function.h:87]: `cast_to` is a static function on `Object`, so we always use it as `Object::cast_to<T>(...)`, not calling it on instances.

**Comment by dalexeev** [comment]
I don't think the linked issue is correct, and thus the solution doesn't look right to me.

File: modules/gdscript/gdscript_function.h (hunk 1 above)
Comment (garychia) [reply modules/gdscript/gdscript_function.h:87]: I see. I was trying to check if an array element type is compatible with another element type.

**Comment by garychia** [comment]
I see. It seems the issue is not about conversions but whether an array has the expected or compatible element type so it can be passed to a function? If so, is it possible (or safe) to implement such a feature?

**Comment by dalexeev** [comment]
This is incorrect. You just removed the check in the static analyzer and GDScript runtime, but didn't solve the "problem", including in the core. The array type is not just the variable type, it is the type of an object in memory (because arrays are passed by reference). This restriction is needed to prevent the following:



```gdscript
var a: Array[Node2D] = []
var b: Array[Node] = a # Now b refers to the same array as a.
b.append(Node.new()) # This is correct for b, but not for a!
print(a[0]) # Node
```

File: modules/gdscript/gdscript_analyzer.cpp [hunk 2]
```diff
@@ -5067,7 +5067,7 @@ bool GDScriptAnalyzer::is_type_compatible(const GDScriptParser::DataType &p_targ
 		if (valid && p_target.builtin_type == Variant::ARRAY && p_source.builtin_type == Variant::ARRAY) {
 			// Check the element type.
 			if (p_target.has_container_element_type() && p_source.has_container_element_type()) {
-				valid = p_target.get_container_element_type() == p_source.get_container_element_type();
+				valid = is_type_compatible(p_target.get_container_element_type(), p_source.get_container_element_type(), false, p_source_node);
```
Comment (dalexeev) [inline modules/gdscript/gdscript_analyzer.cpp:5070]: This is not an error, element types must be equal, supertypes are not supported. I have a [draft](https://github.com/dalexeev/godot/blob/1b596113a112e99fc3769cc9afa6485c81ed1e6c/modules/gdscript/gdscript_analyzer.cpp#L5084-L5092) with the `is_type_compatible()` refactoring in which I added a comment about it.

**Comment by garychia** [comment]
Thanks for the detailed explanation and advice from @dalexeev and others. Looks like this solution is unfeasible. Feel free to tell me what can be done or close this PR if no further discussion or action is needed.

**Comment by dalexeev** [comment]
> Looks like this solution is unfeasible. Feel free to tell me what can be done or close this PR if no further discussion or action is needed.

I think this can't be implemented in a simple way, unfortunately. In theory, we could allow compatible types to be read-only referenced if we added the `write`/`read` attributes to the type system, see https://github.com/godotengine/godot-proposals/discussions/7364#discussioncomment-6576975. But I think that first we should reach a consensus in the discussion before opening a PR.

**Comment by dalexeev** [comment]
We definitely have issues with typed arrays that need to be resolved after we reach a consensus in the [discussion](https://github.com/godotengine/godot-proposals/discussions/7364).

Unfortunately, this PR doesn't look correct. Probably the correct solution to the #80586 problem would be [`read` parameter](https://github.com/godotengine/godot-proposals/discussions/7364#discussioncomment-6576975), [typed array constructor](https://github.com/godotengine/godot/pull/71336) or something else.

Thanks for the contribution nonetheless!

**Pull request closed without merging**